tests/conftest.py

```diff
@@ -56,6 +58,14 @@ def unclosed_files():
     assert _open_files() == fds
 
 
+@pytest.fixture(autouse=True)
+def unclosed_threads():
+    with thread_leaks.assert_none():
+        yield
+        # HAX: close all "singleton producers" before checking thread leaks
+        SingletonProducer._shutdown_all()
```

Comment:
I think it could make sense to make test authors do this:

```
def test_kafka():
    with SingletonProducer.shutdown():
        ...
```

then you don't have to do it here

at the same time, `_shutdown_all` is incredibly cheap if there are no producers, so it makes sense to run it unconditionally

(and i personally like to not have too much indentation in tests